### DISCUSSION TITLE
MorphTo::defaultResource() doesn't select default resource class

### DISCUSSION THREAD

**Discussion by wimski** [laravel/nova-issues#5953] [Ideas & Feature Requests]
- Laravel Version: v10.13.5
- Nova Version: 4.26.5
- PHP Version: 8.2.3

### Description:

If I have a resource with a `MorphTo` field on which I call `->defaultResource(SomeRelatedResource::class)`. I expect - based on the `MorphTo` code below - that the `type` sub field is pre-filled with a default value of the given class.

<details>
<summary>MorphTo field class extract</summary>

```php
    namespace Laravel\Nova\Fields;
    
    class MorphTo
    {
        /**
         * The default related class value for the field.
         *
         * @var (\Closure(\Laravel\Nova\Http\Requests\NovaRequest):(class-string<\Laravel\Nova\Resource>))|class-string<\Laravel\Nova\Resource>
         */
        public $defaultResourceCallable;
    
        /**
         * Set the default relation resource class to be selected.
         *
         * @param  (\Closure(\Laravel\Nova\Http\Requests\NovaRequest):(class-string<\Laravel\Nova\Resource>))|class-string<\Laravel\Nova\Resource>  $resourceClass
         * @return $this
         */
        public function defaultResource($resourceClass)
        {
            $this->defaultResourceCallable = $resourceClass;
    
            return $this;
        }
    
        /**
         * Resolve the default resource class for the field.
         *
         * @param  \Laravel\Nova\Http\Requests\NovaRequest  $request
         * @return string|void
         */
        protected function resolveDefaultResource(NovaRequest $request)
        {
            if ($request->isCreateOrAttachRequest() || $request->isResourceIndexRequest() || $request->isActionRequest()) {
                if (is_null($this->value) && $this->defaultResourceCallable instanceof Closure) {
                    $class = call_user_func($this->defaultResourceCallable, $request);
                } else {
                    $class = $this->defaultResourceCallable;
                }
    
                if (! empty($class) && class_exists($class)) {
                    return $class::uriKey();
                }
            }
        }
    
        /**
         * Prepare the field for JSON serialization.
         *
         * @return array<string, mixed>
         */
        public function jsonSerialize(): array
        {
            $resourceClass = $this->resourceClass;
    
            return with(app(NovaRequest::class), function ($request) use ($resourceClass) {
                $viewable = ! is_null($this->viewable)
                        ? $this->viewable
                        : (! is_null($resourceClass) ? $resourceClass::authorizedToViewAny($request) : true);
    
                return array_merge([
                    // ...
                    'defaultResource' => $this->resolveDefaultResource($request),
                ], parent::jsonSerialize());
            });
        }
    }
```
</details>

The value `defaultResource` is correctly passed to the front-end, but it seems it's not used properly. This is probably due to the following:

```js
// MorphToField.vue

      if (this.shouldSelectInitialResource) {
        if (!this.resourceType && this.field.defaultResource) {
          this.resourceType = this.field.defaultResource
        }
       // ...
      }

    /**
     * Determine if we should select an initial resource when mounting this field
     */
    shouldSelectInitialResource() {
      return Boolean(
        this.editingExistingResource ||
        this.viaRelatedResource ||
        Boolean(this.field.value && this.field.defaultResource)
      )
    },
```

Judging from that last boolean statement, the default resource is only used when there is also a value present. In that case it can never be used as an actual 'default' resource.

**Comment by wimski** [comment]
@crynobone How is this not a bug, but an 'idea'/'feature request'?

`Laravel\Nova\Fields\MorphTo`
> 119: The _**default**_ related _**class**_ value for the field.
>
> 594: Set the _**default**_ relation _**resource class**_ to be selected.
>
> 607: Resolve the _**default resource class**_ for the field.

**Comment by crynobone** [comment]
You request to change the current structure from using defaultResource +
default to only use defaultResource. How is that a bug?

**Reply by wimski** [reply]
Since everything surrounding the 'default resource' setting in the `MorphTo` class says specifically _"default resource **CLASS**"_, I'd except it to use the given class as the default resource class, regardless of any (id) value being present. There's no mention of 'this only works in conjunction with the default method'.

So either this is a bug (behaves differently than described) or the annotations are incorrect/misleading (also kind of a bug). In any case, the documentation could benefit greatly from a section describing how to use defaults for MorphTo fields.

**Comment by crynobone** [comment]
<img width="1303" alt="CleanShot 2023-10-11 at 16 37 34@2x" src="https://github.com/laravel/nova-issues/assets/172966/cf3a47f4-9391-417c-94a9-a4cdbe2107fe">

From the original PR as reference.

**Reply by wimski** [reply]
That clears things up quite a bit. Why isn't this explanation part of the documentation?

In that case my original question stands as a new feature request :slightly_smiling_face:

**Reply by bradley-varol** [reply]
I was also quite confused by this - the behaviour I expected was that only using `defaultResource()` would pre-populate the "commentable_type" field but not the "commentable_id" field.

**Reply by amashkarin** [reply]
I found hack, for apply desired behavior
```
MorphTo::make('commentable')
  ->defaultResource(Post::class)
  ->default(-1) // hack for applying defaultResource
```
